sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -1175,6 +1205,27 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     client.listFunctions(db, pattern)
   }
 
+  /** Detect whether a table is stored with Hive-compatible metadata. */
+  private def isHiveCompatible(table: CatalogTable): Boolean = {
```

Comment:
> So far, the safest way to check the compatibility is to compare the schema.

That is not enough, as I've tried to explain several times. `alterTableSchema` is broken in older Spark versions and can end up creating corrupt tables (where the schema does not match the storage descriptor, for example). So you need a reliable way of detecting compatibility, and the schema is not it.

The closest we have currently is the storage descriptor (the fallback case in my code).

Let me think about what happens when an older Spark touches the tables with the new property set. `alterTableSchema` even in older Spark versions will maintain that information, but maybe other code paths won't.